File: Entities/EasyProgressBar/ProgressBarDefiner.cs
using Celeste.Mod.Entities;
using CorreWithCare.Core;
using CorreWithCare.Utils;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using CorreWithCare.Core.ProgressBar;
using static CorreWithCare.Core.ExtendedAttributes;

namespace CorreWithCare.Entities.EasyProgressBar;

/// <summary>
/// 进度条定义：声明一个进度条的样式，并把它登记到当前会话中。
/// 同一名称与计数器类型的定义会覆盖之前的设置。
/// </summary>
[WorkInProgress]
[CustomEntity("CorreWithCare/ProgressBarDefiner")]
public class ProgressBarDefiner : BaseEntity
{
    public ProgressBarDefiner(EntityData data, vec2 offset) : base(data, offset)
    {
        int direction = data.Int("direction", BarDirection.FromTop);
        bool vertical = BarDirection.IsVertical(direction);

        ProgressBarSettings setting = new()
        {
            Name = data.Attr("sessionName", "default"),
            Direction = direction,
            MoveDuration = data.Float("moveDuration", 1f).Abs(),
            TargetOffset = data.Float("targetOffset", vertical ? 192f : cons.ScreenHeight * 0.1f),
            Border = (data.Float("sessionBorder1", 0),
                    data.Float("sessionBorder2", 100)),
            Size = data.Float("size", vertical ? cons.ScreenHeight * 0.8f : cons.ScreenWidth * 0.8f),
            FontScale = data.Vector2("fontSizeX", "fontSizeY", vec2.One).Abs(),
            BarThickness = data.Float("barSize", 10f).ClampMin(1f),
            FontColor = data.GetCorreColor("fontColor", ccolor.White),
            BarColor = data.GetCorreColor("barColor", ccolor.White),
            GapSize = data.Vector2("gapX", "gapY", vec2.One * 10f).Abs(),
            ShadowShift = data.Vector2("shadowShiftX", "shadowShiftY", vec2.UnitY * 8f),
            TitleName = data.Attr("titleName"),
            IsCounter = data.Bool("isCounter", true),
            TitleScale = data.Vector2("titleScaleX", "titleScaleY", vec2.One),
            TitleColor = data.GetCorreColor("titleColor", ccolor.White),
            Flag = data.Attr("flag"),
            TitleOnBottom = data.Bool("titleOnBottom", true)
        };

        int n = md.Session.ProgressBars.FindIndex(s =>
            s.Name == setting.Name &&
            s is ProgressBarSettings);
        if (n == -1)
        {
            CWCModule.Session.ProgressBars.Add(setting);
        }
        else
        {
            CWCModule.Session.ProgressBars[n] = setting;
        }
    }
}

File: ModFolder/Loenn/entities/ProgressBarDefiner.lua
local drawableSprite = require("structs.drawable_sprite")

local marker = {}

marker.name = "CorreWithCare/ProgressBarDefiner"

marker.placements = {
    name = "marker",
    data = {
        isCounter = true,
        sessionName = "default",
        direction = 0,
        moveDuration = 1,
        targetOffset = 108,
        sessionBorder1 = 0,
        sessionBorder2 = 100,
        size = 1536,
        fontSizeX = 1,
        fontSizeY = 1,
        barSize = 10,
        fontColor = "ffffffff",
        barColor = "ffffffff",
        gapX = 10,
        gapY = 10,
        shadowShiftX = 0,
        shadowShiftY = 8,
        titleName = "",
        titleScaleX = 1,
        titleScaleY = 1,
        titleColor = "ffffffff",
        flag = "",
        titleOnBottom = true,
    }
}

marker.fieldInformation = {
    fontColor = {
        fieldType = "color",
        useAlpha = true,
    },
    barColor = {
        fieldType = "color",
        useAlpha = true,
    },
    titleColor = {
        fieldType = "color",
        useAlpha = true,
    },
    -- 0/1 为水平进度条（从上/下方进出），2/3 为垂直进度条（从左/右侧进出）
    direction = {
        fieldType = "integer",
        options = {
            ["From Top"] = 0,
            ["From Bottom"] = 1,
            ["From Left"] = 2,
            ["From Right"] = 3,
        },
        editable = false,
    },
}

function marker.sprite(room, entity)
    local sprite = {}
    local iconSprite = drawableSprite.fromTexture("CorreWithCare/LoennIcons/Marker", entity)

    table.insert(sprite, iconSprite)
    return sprite
end

return marker
